### ISSUE TITLE
A Botton part of big digram can be hidden.

### ISSUE THREAD

**Issue by satabol** [mike-lischke/vscode-antlr4#243]
Proposal.

**Problem:**

Windows 11,  vscode 1.86.0, vscode-antlr4 **2.4.6**.

Long rule:

https://github.com/mike-lischke/vscode-antlr4/assets/14288520/6201a106-15d6-4c2b-a800-3d58f166f653

If append this css code then all looks good:
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/222d561d-f10e-46ca-b2f1-4bf732a753a7)

```
#container svg:last-child {
    width: 100%;
    margin-bottom: 50px;
}
```

https://github.com/mike-lischke/vscode-antlr4/assets/14288520/7a9450fc-5cbd-4c1a-9dcc-26cde462a142

**Comment by mike-lischke** [comment]
IIRC this is not only a problem for the last rule. I have seen this also for rules in the middle somewhere. Last time I looked it was so that the railroad library did not correctly compute the total height of the diagram. Needs investigation, but your proposal would only fix a part of the problem.

**Comment by satabol** [comment]
Do you have an example about rules in the middle somewhere? If not can you put them later?

**Comment by satabol** [comment]
I found something interesting. You build tab with next structure.

1. semitransparent header with g4 info:
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/f8dab7c6-c210-4e7d-956f-8ab7a608608f)

2. You have a container with rail structure:
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/004d1b1d-7e13-4694-a83a-3d8cc9c9e061)

In 2 you use a property with name "padding" that has a parameter "60px". This has incorrect influence to "height:100vh" (100vh do not enlarge on 60px):
(example: https://css-tricks.com/almanac/properties/p/padding/#aa-padding-and-element-width-calculations)

![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/e62efc12-1174-4b9e-8ab1-92a168a32b74)

To fix this I found a property "**box-sizing: border-box**":
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/9775cf0f-23f8-44be-a1bc-9b1efa80f0ac)

Now all looks good: first element is not under a header and a last element is not hidden. What do you think?

info of a box-sizing:

- https://www.w3.org/TR/css-sizing-3/#box-sizing
- https://caniuse.com/css3-boxsizing

**Comment by satabol** [comment]
> that the railroad library did not correctly compute the total height of the diagram

Now I think that this railroad issue is another issue.

**Comment by mike-lischke** [comment]
My assumption was that the missing space at the end is the same problem like missing space in the middle of the document:

![Bildschirmfoto 2024-02-13 um 09 30 01](https://github.com/mike-lischke/vscode-antlr4/assets/2398704/5531df0d-04f4-4ae8-a5c4-5efa7e5f5a32)

But you might have a point in your analysis of other spacing and the boxing model. I'll have to investigate.

**Comment by satabol** [comment]
Can you give me this example? Or send it by email satabolxyz@gmail.com ? I promise not to show anyone.

**Comment by mike-lischke** [comment]
It's [open source](https://github.com/mysql/mysql-shell-plugins/blob/master/gui/frontend/src/parsing/mysql/MySQLParser.g4) like everything I do on Github 😄

**Comment by satabol** [comment]
I'm sorry I found MySQLParser.g4 file with another syntax. Now I use your grammar. Can you show your extension settings?

**Comment by mike-lischke** [comment]
It probably doesn't make a difference, but here they are:

```json
    "antlr4.rrd.saveDir": "/Users/Mike/Desktop/rrd",
    "antlr4.atn.saveDir": "",
    "antlr4.atn.maxLabelCount": 10,
    "antlr4.call-graph.saveDir": "/tmp/",
    "antlr4.debug": {
        "visualParseTreeHorizontal": true,
        "visualParseTreeClustered": true
    },
    "antlr4.generation": {
        "mode": "internal",
        "listeners": true,
        "visitors": false,
        "outputDir": "./generated"
    },
    "antlr4.format": {
        "disabled": false,
        "alignTrailingComments": true,
        "allowShortBlocksOnASingleLine": true,
        "breakBeforeBraces": false,
        "columnLimit": 150,
        "continuationIndentWidth": 4,
        "indentWidth": 4,
        "keepEmptyLinesAtTheStartOfBlocks": false,
        "maxEmptyLinesToKeep": 1,
        "reflowComments": false,
        "spaceBeforeAssignmentOperators": true,
        "tabWidth": 4,
        "useTab": false,
        "alignColons": "none",
        "singleLineOverrulesHangingColon": false,
        "allowShortRulesOnASingleLine": false,
        "alignSemicolons": "ownLine",
        "breakBeforeParens": false,
        "ruleInternalsOnSingleLine": false,
        "minEmptyLines": 1,
        "groupedAlignments": true,
        "alignFirstTokens": true,
        "alignLexerCommands": false,
        "alignActions": false,
        "alignLabels": true,
        "alignTrailers": false
    },
```

**Comment by satabol** [comment]
At last I got these errors after settings:
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/a84f68e0-60e7-4a25-a749-dc7afbdaa411)

Nice:
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/5bd8455a-bd54-4d3c-ae4c-720e272ed329)

)))

**Comment by mike-lischke** [comment]
Yes, that's the setting for wrapping long sequence lines in the diagram. Of course, if no wrapping occurs then the display is fine. IMO the RRD script does not fully cover some of the cases we use in these diagrams.

**Comment by satabol** [comment]
I found something - version of railroad.js file in this repo is too old. I try last (newest) version and all look good:

https://github.com/tabatkins/railroad-diagrams/blob/gh-pages/railroad.js
![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/115743a7-488c-4f5f-81c3-47306fc6ce14)

and all look good

![image](https://github.com/mike-lischke/vscode-antlr4/assets/14288520/01883505-3557-4045-b50c-3192bdca4c0e)

I use setting "wrapAfter"=60.

**Comment by mike-lischke** [comment]
That would be great news! [This](https://github.com/tabatkins/railroad-diagrams/issues/106) just fixed issue looks very much like the problems here. Unfortunately, this project neither releases packages nor bundles, so it's tricky to know when a new release comes (if they even make releases).

**Comment by satabol** [comment]
Happy to help you. I just compare count of lines of code. Intuition. ))) Are you planning to update this library?

**Comment by mike-lischke** [comment]
Yes, I will update the library, but I'm pretty busy with other stuff. Need to find some time for it...
